python/cudf/cudf/utils/utils.py

```diff
@@ -489,39 +489,64 @@ def get_appropriate_dispatched_func(
         if cupy_func is func:
             return NotImplemented
 
-        cupy_compatible_args = get_cupy_compatible_args(args)
-        cupy_output = cupy_func(*cupy_compatible_args, **kwargs)
-        return cast_to_appropriate_cudf_type(cupy_output)
+        cupy_compatible_args, index = _get_cupy_compatible_args_index(args)
+        if cupy_compatible_args:
+            cupy_output = cupy_func(*cupy_compatible_args, **kwargs)
+            return _cast_to_appropriate_cudf_type(cupy_output, index)
+        else:
+            return NotImplemented
     else:
         return NotImplemented
 
 
-def cast_to_appropriate_cudf_type(val):
+def _cast_to_appropriate_cudf_type(val, index=None):
     # TODO Handle scalar
     if val.ndim == 0:
         return cudf.Scalar(val).value
     # 1D array
     elif (val.ndim == 1) or (val.ndim == 2 and val.shape[1] == 1):
-        return cudf.Series(val)
+        if (index is None) or (len(index) == len(val)):
+            return cudf.Series(val, index=index)
+        else:
+            # if index is not None and is of a different length
+            # than the index cupy dispatching behaviour is undefined
+            return NotImplemented
     else:
         return NotImplemented
 
 
-def get_cupy_compatible_args(args):
+def _get_cupy_compatible_args_index(args, ser_index=None):
+    """
+     This function returns cupy compatible arguments and output index
+     if conversion is not possible it returns None
+    """
+
     casted_ls = []
     for arg in args:
         if isinstance(arg, cp.ndarray):
             casted_ls.append(arg)
         elif isinstance(arg, cudf.Series):
-            casted_ls.append(arg.values)
+            if _are_indexes_alligned(ser_index, arg.index):
+                ser_index = arg.index
+                casted_ls.append(arg.values)
+            else:
+                # dont dispatch if the index does not line up
+                return None, ser_index
         elif isinstance(arg, Sequence):
-            # handle list of inputs for functions like
```

Comment:
We can handle this as we were previously but I could not find a use case where we dispatch to `cupy` with list inputs and our result index can be aligned without ambiguity. I am not sure what the best way to handle this is but I think currently not supporting them should be fine.

Reply:
This is for cases like `np.some_api([cudf_series_1, cudf_series_2, cudf_series_3])` right, e.g. concatenate? This seems to give back a numpy array in the case of pandas. We'll return `NotImplemented` in this case so what will happen? Will `__array__` end up being called and we'd get a NumPy array out of this?

Reply:
So we currently do throw `TypeError` non implemented error if this happens. This is the following https://numpy.org/neps/nep-0018-array-function-protocol.html  which recommends we return the  return the sentinel value `NotImplemented`, indicating that the function is not implemented by these types.  

I have added a test cast to ensure we cover that case here.  

https://github.com/rapidsai/cudf/blob/7a12226ee6127e57b7e5671ed21dec2cf5278c06/python/cudf/cudf/tests/test_array_function.py#L121-L131

Code example of the same. 
```python
>>> s = cudf.Series([1,2,3])
>>> np.concatenate([s,s,s])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<__array_function__ internals>", line 6, in concatenate
TypeError: no implementation found for 'numpy.concatenate' on types that implement __array_function__: [<class 'cudf.core.series.Series'>]
```
```

Please let me know if i need to address anything else, @brandon-b-miller .

Reply:
That makes sense - but the operation seems to work on pandas series objects and returns a numpy array:

```
>>> import pandas as pd
>>> s = pd.Series([1,2,3])
>>> import numpy as np
>>> np.concatenate([s,s,s])
array([1, 2, 3, 1, 2, 3, 1, 2, 3])
```

although I suspect that this is calling `__array__` somewhere, which I guess we would NOT want to do. So I think the code here is fine - I am just wondering if there are any implications to differing from pandas here.

Reply:
We can probably add these functions like ( `concatenate`) to the `cudf`/`cudf.Series` namespace if we need to support it but I don't expect a lot of these to be out there in the wild honestly and tackle them as we run into it, as we error here **loudly** these cases should be easy to track.  

Also from my exploration running stuff like `np.concatenate` and `np.vstack` on a series object does not really make sense from a user perspective too.  
 

The other alternative is handling them with `cupy` dispatch which seems off to me.  We can probably tackle this in follow up as we run into issues.